app/tx_cache.go

```diff
@@ -0,0 +1,74 @@
+package app
+
+import (
+	"crypto/sha256"
+	"sync"
+	"time"
+)
+
+// TxValidationCache caches the results of expensive transaction validation
+// to avoid repeating the same work in ProcessProposal that was already done in CheckTx.
+type TxValidationCache struct {
+	mu    sync.RWMutex
+	cache map[string]*bool
+}
+
+// NewTxValidationCache creates a new transaction validation cache
+func NewTxValidationCache(ttl time.Duration) *TxValidationCache {
+	return &TxValidationCache{
+		cache: make(map[string]*bool),
+	}
+}
+
+// getTxKey generates a deterministic key for a transaction
+func (c *TxValidationCache) getTxKey(tx []byte) string {
+	hash := sha256.Sum256(tx)
+	return string(hash[:])
+}
+
+// Get retrieves a validation result from the cache
+func (c *TxValidationCache) Get(tx []byte) bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	key := c.getTxKey(tx)
+	result, exists := c.cache[key]
+	if !exists {
+		return false
+	}
+
+	return *result
+}
+
+// Set stores a validation result in the cache
+func (c *TxValidationCache) Set(tx []byte, result bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	key := c.getTxKey(tx)
+	c.cache[key] = &result
+}
+
+// Clear removes all entries from the cache
+func (c *TxValidationCache) Clear() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cache = make(map[string]*bool)
+}
+
+// Cleanup removes expired entries from the cache
+func (c *TxValidationCache) Cleanup() {
```

Comment:
I think we should add a method that only deletes certain keys. Also, I think it would be useful to benchmark how much time it takes to cleanup ~500 entry from the map. This would give us an idea of the added cost when running finalize block. We might even want to run the cleanup in parallel to finalize block.

Also, given the cleanup will take a certain period of time, it might block CheckTx and have an impact on the mempool since checkTx will take way longer during the deletion phase. I'm not sure which is the best solution but a benchmark of how much time it takes to delete entries would be useful

Reply:
I think because we might not include all txs in a block and want the cache for next block.

I would love to see some benchmarks too.